src/validation.cpp

```diff
@@ -5225,6 +5225,33 @@ ChainstateManager::~ChainstateManager()
     }
 }
 
+bool ChainstateManager::IsBlockInMainOrBestChain(const uint256& blockhash) const
+{
+    AssertLockHeld(cs_main);
+
+    const CBlockIndex* index{m_blockman.LookupBlockIndex(blockhash)};
+    if (!index) return false;
+
+    if (m_best_header->GetAncestor(index->nHeight) == index) return true;
```

Comment:
This sounds good to me, however I am not sure if it should replace the `m_best_header` ancestor check or if it should be added as a third check.

Ancestors of `m_best_header` (or any header that has more work than the active tip) could have less work than the active tip and therefore simply doing the "more work" check would lead to `IsBlockInMainOrBestChain` returning false for them. Doing the ancestor check for all headers that have more work than the tip would be ideal but that does not seem easily doable atm.

Maybe we should use `BlockRequestAllowed` instead and get rid of `IsBlockInMainOrBestChain`. What do you think of that?